src/miniscript/iter.rs

```diff
@@ -98,43 +94,93 @@ impl<Pk: MiniscriptKey, Ctx: ScriptContext> Miniscript<Pk, Ctx> {
+        match self.node {
+            Terminal::PkH(ref hash) => vec![PkPkh::HashedPubkey(hash.clone())],
+            Terminal::PkK(ref key) => vec![PkPkh::PlainPubkey(key.clone())],
+            Terminal::Multi(_, ref keys) => keys
+                .into_iter()
+                .map(|key| PkPkh::PlainPubkey(key.clone()))
+                .collect(),
             _ => vec![],
         }
     }
+
+    /// Returns `Option::Some` with cloned n'th public key from the current miniscript item,
+    /// if any. Otherwise returns `Option::None`.
+    ///
+    /// NB: The function analyzes only single miniscript item and not any of its descendants in AST.
+    pub fn get_nth_pk(&self, n: usize) -> Option<Pk> {
+        match (&self.node, n) {
+            (&Terminal::PkK(ref key), 0) => Some(key.clone()),
+            (&Terminal::Multi(_, ref keys), _) => keys.get(n).cloned(),
+            _ => None,
+        }
+    }
+
+    /// Returns `Option::Some` with hash of n'th public key from the current miniscript item,
+    /// if any. Otherwise returns `Option::None`.
+    ///
+    /// For each public key the function computes hash; for each hash of the public key the function
+    /// returns it cloned copy.
+    ///
+    /// NB: The function analyzes only single miniscript item and not any of its descendants in AST.
+    pub fn get_nth_pkh(&self, n: usize) -> Option<Pk::Hash> {
+        match (&self.node, n) {
+            (&Terminal::PkH(ref hash), 0) => Some(hash.clone()),
+            (&Terminal::PkK(ref key), 0) => Some(key.to_pubkeyhash()),
+            (&Terminal::Multi(_, ref keys), _) => keys.get(n).map(Pk::to_pubkeyhash),
+            _ => None,
+        }
+    }
+
+    /// Returns `Option::Some` with hash of n'th public key or hash from the current miniscript item,
+    /// if any. Otherwise returns `Option::None`.
+    ///
+    /// NB: The function analyzes only single miniscript item and not any of its descendants in AST.
+    pub fn get_nth_pk_pkh(&self, n: usize) -> Option<PkPkh<Pk>> {
+        match (&self.node, n) {
+            (&Terminal::PkH(ref hash), 0) => Some(PkPkh::HashedPubkey(hash.clone())),
+            (&Terminal::PkK(ref key), 0) => Some(PkPkh::PlainPubkey(key.clone())),
+            (&Terminal::Multi(_, ref keys), _) => {
+                keys.get(n).map(|key| PkPkh::PlainPubkey(key.clone()))
+            }
+            _ => None,
+        }
+    }
 }
 
 /// Iterator for traversing all [Miniscript] miniscript AST references starting from some specific
 /// node which constructs the iterator via [Miniscript::iter] method.
 pub struct Iter<'a, Pk: 'a + MiniscriptKey, Ctx: 'a + ScriptContext> {
     next: Option<&'a Miniscript<Pk, Ctx>>,
-    path: Vec<(&'a Miniscript<Pk, Ctx>, usize)>,
+    // Here we store vec of path elements, where each element is a tuple, consisting of:
+    // 1. Miniscript node on the path
+    // 2. It's branches stored as a vec (used for avoiding multiple vec allocations
```

Comment:
wrong "It's"

Reply:
Also there is no need to do **any** `Vec` allocations here

Reply:
No allocations now. Also, removed `branches` fn as you asked.